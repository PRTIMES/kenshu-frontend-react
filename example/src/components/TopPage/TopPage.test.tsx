import { PropsWithChildren, Suspense } from 'react';
import { TopPage } from '.';
import { act, render, waitFor, within } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { setupMockServer } from '../../__tests__/msw';
import { rest } from 'msw';
import userEvent from '@testing-library/user-event';

const mockHandlers = [
  rest.get('http://localhost:8000/api/tasks', (_, res, ctx) =>
    res(
      ctx.json({
        tasks: [
          {
            id: '1',
            title: 'Get up',
            createdAt: '2021-01-01T00:00:00.000Z',
            finishedAt: '2021-01-01T00:00:00.000Z',
          },
          {
            id: '2',
            title: 'Eat breakfast',
            createdAt: '2021-01-01T00:00:00.000Z',
            finishedAt: null,
          },
          {
            id: '3',
            title: 'Go to school',
            createdAt: '2021-01-01T00:00:00.000Z',
            finishedAt: null,
          },
        ],
      }),
    ),
  ),
];

const mockServer = setupMockServer(mockHandlers);

const wrapper = ({ children }: PropsWithChildren) => (
  <QueryClientProvider client={new QueryClient()}>
    <Suspense fallback={<p>loading</p>}>{children}</Suspense>
  </QueryClientProvider>
);

describe('TopPage', () => {
  beforeAll(() => void mockServer.listen());
  afterEach(() => void mockServer.resetHandlers());
  afterAll(() => void mockServer.close());

  it('HTML とスタイルが変わらないこと', async () => {
    const { queryByText, asFragment } = render(<TopPage />, { wrapper });
    await waitFor(() => expect(queryByText('loading')).toBeNull());

    expect(asFragment()).toMatchSnapshot();
  });

  it('タスクが表示されていること', async () => {
    const { queryByText, getByRole } = render(<TopPage />, {
      wrapper,
    });
    await waitFor(() => expect(queryByText('loading')).toBeNull());

    expect(getByRole('list')).toBeDefined();
    expect(within(getByRole('list')).getByText('Get up')).toBeDefined();
  });

  it('タスクの作成ボタンが表示されていること', async () => {
    const { queryByText, getByRole } = render(<TopPage />, { wrapper });
    await waitFor(() => expect(queryByText('loading')).toBeNull());

    expect(getByRole('button', { name: 'タスクを追加' })).toBeDefined();
  });

  it('タスクの削除ボタンが表示されていること', async () => {
    const { queryByText, getAllByRole } = render(<TopPage />, { wrapper });
    await waitFor(() => expect(queryByText('loading')).toBeNull());

    expect(getAllByRole('button', { name: '削除' })).toHaveLength(3);
  });

  it('タスクのタイトルをクリックすると編集フォームが表示されること', async () => {
    const { queryByText, getByRole } = render(<TopPage />, { wrapper });
    await waitFor(() => expect(queryByText('loading')).toBeNull());

    await act(() =>
      userEvent.click(
        getByRole('button', { name: 'Get up のタイトルを編集する' }),
      ),
    );
    expect(getByRole('textbox')).toHaveValue('Get up');
  });
});
