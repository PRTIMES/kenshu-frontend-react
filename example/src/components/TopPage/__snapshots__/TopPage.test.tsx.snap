// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TopPage > HTML とスタイルが変わらないこと 1`] = `
<DocumentFragment>
  <main
    class="min-h-screen bg-stone-50 px-2 pt-8"
  >
    <div
      class="mx-auto flex max-w-3xl flex-col gap-4 rounded-lg bg-white p-5 shadow-md"
    >
      <header
        class="flex items-center justify-between"
      >
        <h1
          class="text-lg font-bold"
        >
          TODO リスト
        </h1>
        <button
          class="rounded-lg bg-orange-500 px-2 py-1 text-sm text-white hover:bg-orange-400"
          type="button"
        >
          タスクを追加
        </button>
      </header>
      <ul
        class="grid list-none gap-4"
      >
        <li
          class="flex items-center justify-between"
        >
          <div
            class="flex items-center gap-4"
          >
            <button
              aria-label="未完了に戻す"
              class="h-6 w-6 rounded-md border-2 border-stone-500 bg-stone-500"
              type="button"
            />
            <button
              class="text-stone-400 line-through"
            >
              <span>
                Get up
              </span>
              <span
                class="sr-only"
              >
                のタイトルを編集する
              </span>
            </button>
          </div>
          <button
            class="rounded-lg bg-red-500 px-2 py-1 text-sm text-white hover:bg-red-400"
            type="button"
          >
            削除
          </button>
        </li>
        <li
          class="flex items-center justify-between"
        >
          <div
            class="flex items-center gap-4"
          >
            <button
              aria-label="完了する"
              class="h-6 w-6 rounded-md border-2 border-stone-500"
              type="button"
            />
            <button
              class=""
            >
              <span>
                Eat breakfast
              </span>
              <span
                class="sr-only"
              >
                のタイトルを編集する
              </span>
            </button>
          </div>
          <button
            class="rounded-lg bg-red-500 px-2 py-1 text-sm text-white hover:bg-red-400"
            type="button"
          >
            削除
          </button>
        </li>
        <li
          class="flex items-center justify-between"
        >
          <div
            class="flex items-center gap-4"
          >
            <button
              aria-label="完了する"
              class="h-6 w-6 rounded-md border-2 border-stone-500"
              type="button"
            />
            <button
              class=""
            >
              <span>
                Go to school
              </span>
              <span
                class="sr-only"
              >
                のタイトルを編集する
              </span>
            </button>
          </div>
          <button
            class="rounded-lg bg-red-500 px-2 py-1 text-sm text-white hover:bg-red-400"
            type="button"
          >
            削除
          </button>
        </li>
      </ul>
    </div>
  </main>
</DocumentFragment>
`;
